x86_64: Make sure to clone ap list in vasprintf

btashton opened the pull request:
## Summary
VA_LIST is getting clobbered because it is a pointer to a structure
on the stack and we must traverse it twice.  Clone it like we do
for x86_32

This was first noticed using the `ps` command.
## Impact
Correct output is now generated by the `vasprintf` function on x86_64.

## Testing
From NSH I was able to run `ps` and not have the print output corrupted.

xiaoxiang781216 commented on libs/libc/stdio/lib_vasprintf.c:
```diff
@@ -58,7 +58,7 @@
 #define ap2 ap
 
-#if defined(CONFIG_ARCH_X86)
+#if defined(CONFIG_ARCH_X86) || defined(CONFIG_ARCH_X86_64)
```
Should we change like vsyslog to avoid the same problem happen on the new arch:
#ifdef va_copy
      va_list copy;

      va_copy(copy, ap);
      nx_vsyslog(priority, fmt, &copy);
      va_end(copy);
#else
      nx_vsyslog(priority, fmt, &ap);
#endif

btashton replied:
I think it is OK.  If anything this is a place where we should add the conditional to avoid the copy if the arch is not leaving it on the stack.

From `stdarg.h` in gcc 
```
#if !defined(__STRICT_ANSI__) || __STDC_VERSION__ + 0 >= 199900L || defined(__GXX_EXPERIMENTAL_CXX0X__)
#define va_copy(d,s)        __builtin_va_copy(d,s)
#endif
#define __va_copy(d,s)        __builtin_va_copy(d,s)
```

and `stdarg.h` in clang
```
/* GCC always defines __va_copy, but does not define va_copy unless in c99 mode
 * or -ansi is not specified, since it was not part of C90.
 */
#define __va_copy(d,s) __builtin_va_copy(d,s)

#if __STDC_VERSION__ >= 199901L || __cplusplus >= 201103L || !defined(__STRICT_ANSI__)
#define va_copy(dest, src)  __builtin_va_copy(dest, src)
#endif
```

patacongo replied:
We do need to do something differently.  Architectural dependencies or conditional logic  are not permitted in common code.  We control all architecture-specific logic under libs/linc/machine or in header files at include/arch.

btashton replied:
Is there a downside to just always calling the copy as is done with vsyslog? I'm not sure how to avoid the stack usage without having machine specific code in the function.

xiaoxiang781216 replied:
It isn't a big issue because va_copy is a simple pointer assignment for no x64 like platform.

patacongo commented:
I am wondering why we do not see these issues on the x86_64-based simulation.  Same architecture, same tools, real hardware.  The ostest runs with no failures on that platform.  What is different about the qemu platform that causes code changes?

btashton commented:
Does the ps command work on the sim when compiled with x86_64?  That is where I hit this.

patacongo commented:
I have used it many times with the simulation.  In fact, all OS features are developed on the simulation and the sim:ostest configuration runs with no problems.

There is the sim:nsh configuration but is a little awkard because it has a login and enables binfs.  Try it.  The user name is admin and the passward is Adminstrator (see boards/sim/sim/sim/README.txt if I got that wrong.

The binfs means that you have to mount like:

    nsh> mount -t binfs bin

Before you can execute built-in program.

I build it a re-verify when I get the chance.

btashton commented:
@patacongo 
It looks like we do check for x86_64 in the sim 

#elif defined(CONFIG_ARCH_SIM) && (defined(CONFIG_HOST_X86) || defined(CONFIG_HOST_X86_64))
# define CLONE_APLIST 1
# undef ap2
#endif

xiaoxiang781216 commented:
> I am wondering why we do not see these issues on the x86_64-based simulation. Same architecture, same tools, real hardware. The ostest runs with no failures on that platform. What is different about the qemu platform that causes code changes?

We hit the similar issue for syslog on x64 simulator, and modify syslog to the current state.

xiaoxiang781216 commented:
> @patacongo
> It looks like we do check for x86_64 in the sim
> 
> ```
> #elif defined(CONFIG_ARCH_SIM) && (defined(CONFIG_HOST_X86) || defined(CONFIG_HOST_X86_64))
> # define CLONE_APLIST 1
> # undef ap2
> #endif
> ```

Yes, that's why I suggest to trigger the copy by checking va_copy existence instead of arch macro.

btashton commented:
> > @patacongo
> > It looks like we do check for x86_64 in the sim
> > 
> > ```
> > #elif defined(CONFIG_ARCH_SIM) && (defined(CONFIG_HOST_X86) || defined(CONFIG_HOST_X86_64))
> > # define CLONE_APLIST 1
> > # undef ap2
> > #endif
> > ```
> 
> Yes, that's why I suggest to trigger the copy by checking va_copy existence instead of arch macro.

Ah I misunderstood. I thought you were asking me to add the arch macro to this code.

I think this might add the copy to architectures where we could avoid copying, but I'll need to take a look at the headers.

btashton commented:
After reading this https://stackoverflow.com/questions/8047362/is-gcc-mishandling-a-pointer-to-a-va-list-passed-to-a-function/8048892#8048892
I am more inclined to just call va_copy if it is defined, trying to do something ugly to determine the underlying type of va_list is probably going to be even worse.

@patacongo does that sound reasonable?

btashton commented:
Ok pushed the change.  I did notice that we are missing a lot of va_end calls. I added it to this function that I changed, but we should probably look at the other functions, especially ones that end early.

The style check error I think is a false positive.

patacongo commented:
> does that sound reasonable?

I am not really fully up to speed on this discussion.  What you say seems reasonable to me.  Let's let @xiaoxiang781216 make the final call because because he has a stake in this.
